### PR TITLE
Issue #8761: Move metadata generation code to main from test folder

### PR DESCRIPTION
**Pull request by gaurabdg**
#8761

### CODE REVIEW

**Comment by romani** [comment]
Please resolve IDEAs violations.

**Comment by romani** [comment]
wercker failure:
```
Command: ./.ci/wercker.sh no-error-checkstyles-sevntu
....
[INFO] --- maven-checkstyle-plugin:3.1.1:check (sevntu-checkstyle-check) @ checkstyle ---
[INFO] There are 20 errors reported by Checkstyle 8.36-SNAPSHOT with /pipeline/source/config/checkstyle_sevntu_checks.xml ruleset.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:[105,37] (extension) MultipleStringLiteralsExtended: The String "Check" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:[106,73] (extension) MultipleStringLiteralsExtended: The String "Check" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaWriter.java:[64,34] (extension) MultipleStringLiteralsExtended: The String "name" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaWriter.java:[69,48] (extension) MultipleStringLiteralsExtended: The String "description" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaWriter.java:[95,39] (extension) MultipleStringLiteralsExtended: The String "name" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaWriter.java:[105,64] (extension) MultipleStringLiteralsExtended: The String "description" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaWriter.java:[116,32] (extension) MultipleStringLiteralsExtended: The String "/" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaWriter.java:[126,44] (extension) MultipleStringLiteralsExtended: The String "checkstyle" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaWriter.java:[126,60] (extension) MultipleStringLiteralsExtended: The String "checkstyle" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaWriter.java:[128,45] (extension) MultipleStringLiteralsExtended: The String "/" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaWriter.java:[129,84] (extension) MultipleStringLiteralsExtended: The String ".xml" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaWriter.java:[136,79] (extension) MultipleStringLiteralsExtended: The String ".xml" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaReader.java:[114,54] (extension) MultipleStringLiteralsExtended: The String "name" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaReader.java:[117,64] (extension) MultipleStringLiteralsExtended: The String "description" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaReader.java:[140,61] (extension) MultipleStringLiteralsExtended: The String "name" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaReader.java:[142,35] (extension) MultipleStringLiteralsExtended: The String "default-value" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaReader.java:[143,73] (extension) MultipleStringLiteralsExtended: The String "default-value" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaReader.java:[145,35] (extension) MultipleStringLiteralsExtended: The String "validation-type" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaReader.java:[146,75] (extension) MultipleStringLiteralsExtended: The String "validation-type" appears 2 times in the file.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaReader.java:[148,71] (extension) MultipleStringLiteralsExtended: The String "description" appears 2 times in the file.
```

please fix.

**Comment by romani** [comment]
All pmd/spotbugs issues has to be resolved https://travis-ci.org/github/checkstyle/checkstyle/jobs/721311223#L653

**Comment by romani** [comment]
please hardcode [coverage numbers](https://travis-ci.org/github/checkstyle/checkstyle/jobs/721311222#L1312) in xml for each class in pom.xml
```
[INFO] --- jacoco-maven-plugin:0.8.5:check (default-check) @ checkstyle ---
1310[INFO] Loading execution data file /home/travis/build/checkstyle/checkstyle/target/jacoco.exec
1311[INFO] Analyzed bundle 'checkstyle' with 557 classes
1312[WARNING] Rule violated for class com.puppycrawl.tools.checkstyle.meta.JavadocMetadataScraper: lines covered ratio is 0.98, but expected minimum is 1.00
1313[WARNING] Rule violated for class com.puppycrawl.tools.checkstyle.meta.JavadocMetadataScraper: branches covered ratio is 0.85, but expected minimum is 1.00
1314[WARNING] Rule violated for class com.puppycrawl.tools.checkstyle.meta.XmlMetaWriter: lines covered ratio is 0.93, but expected minimum is 1.00
1315[WARNING] Rule violated for class com.puppycrawl.tools.checkstyle.meta.XmlMetaWriter: branches covered ratio is 0.80, but expected minimum is 1.00
1316[WARNING] Rule violated for class com.puppycrawl.tools.checkstyle.meta.XmlMetaReader: lines covered ratio is 0.00, but expected minimum is 1.00
1317[WARNING] Rule violated for class com.puppycrawl.tools.checkstyle.meta.XmlMetaReader: branches covered ratio is 0.00, but expected minimum is 1.00
1318[WARNING] Rule violated for class com.puppycrawl.tools.checkstyle.meta.ModuleDetails: lines covered ratio is 0.76, but expected minimum is 1.00
1319[WARNING] Rule violated for class com.puppycrawl.tools.checkstyle.meta.MetadataGenerator: lines covered ratio is 0.96, but expected minimum is 1.00
1320[WARNING] Rule violated for class com.puppycrawl.tools.checkstyle.meta.MetadataGenerator: branches covered ratio is 0.90, but expected minimum is 1.00
```
build should pass

**Comment by romani** [comment]
missed pitest is detected https://travis-ci.org/github/checkstyle/checkstyle/jobs/721311256#L273

pelase setup execution of it and put in ignore list all cases. Example:
https://github.com/checkstyle/checkstyle/blob/52aa25ba1e5b5971f231e97a4a595e61593749a3/.ci/pitest.sh#L69-L72

**Comment by gaurabdg** [comment]
@romani 
I have resolved everything except for pitest. Since all the mutations are generating new XML files, it will take a lot of time to check(tens of hours). 
I will add pitest coverage when I will implement normal UTs where I'll do assertions without writing to files.

**Comment by romani** [comment]
@gaurabdg yes please add all pitest survived item to ignoredItems in shell script, build will pass

**Comment by romani** [comment]
Please resolve https://travis-ci.org/github/checkstyle/checkstyle/builds/721517758

**Comment by strkkk** [comment]
@gaurabdg AppVeyor failure 
```
[INFO] Total bugs: 1
[ERROR] Medium: Nullcheck of files at line 87 of value previously dereferenced in com.puppycrawl.tools.checkstyle.meta.MetadataGenerator.dumpMetadata(Checker, String) [com.puppycrawl.tools.checkstyle.meta.MetadataGenerator, com.puppycrawl.tools.checkstyle.meta.MetadataGenerator] At MetadataGenerator.java:[line 87]Redundant null check at MetadataGenerator.java:[line 94] RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE
```

**Comment by gaurabdg** [comment]
> @gaurabdg yes please add all pitest survived item to ignoredItems in shell script, build will pass

Pitest is failing to run since each time it is writing a new file for each of the mutation, and taking a lot of time to finish, ran for more than an hour yesterday, still didint finish. So, the solutiion is to add simple UTs where it wont be written to files and simple object assertions would be done. I am planning to write these after succesful plugin integration.
Patch used:
```
diff --git a/.circleci/config.yml b/.circleci/config.yml
index 37139dc9a..987ff8b02 100644
--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -64,6 +64,7 @@ jobs:
               echo "./.ci/pitest.sh pitest-common"  >> commands.txt
               echo "./.ci/pitest.sh pitest-imports" >> commands.txt
               echo "./.ci/pitest.sh pitest-ant"     >> commands.txt
+              echo "./.ci/pitest.sh pitest-meta"     >> commands.txt
               CMD="$(circleci tests split commands.txt)"
               echo "Command: $CMD"
               eval $CMD
diff --git a/pom.xml b/pom.xml
index a60f1c190..9b03f174e 100644
--- a/pom.xml
+++ b/pom.xml
@@ -3362,5 +3362,47 @@
         </plugins>
       </build>
     </profile>
+    <profile>
+      <id>pitest-meta</id>
+      <properties>
+        <skipTests>true</skipTests>
+        <jacoco.skip>true</jacoco.skip>
+      </properties>
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>org.pitest</groupId>
+            <artifactId>pitest-maven</artifactId>
+            <version>${pitest.plugin.version}</version>
+            <configuration>
+              <mutators>
+                <mutator>CONDITIONALS_BOUNDARY</mutator>
+                <mutator>CONSTRUCTOR_CALLS</mutator>
+                <mutator>FALSE_RETURNS</mutator>
+                <mutator>INCREMENTS</mutator>
+                <mutator>INVERT_NEGS</mutator>
+                <mutator>MATH</mutator>
+                <mutator>NEGATE_CONDITIONALS</mutator>
+                <mutator>REMOVE_CONDITIONALS</mutator>
+                <mutator>RETURN_VALS</mutator>
+                <mutator>TRUE_RETURNS</mutator>
+                <mutator>VOID_METHOD_CALLS</mutator>
+              </mutators>
+              <targetClasses>
+                <param>com.puppycrawl.tools.checkstyle.meta.*</param>
+              </targetClasses>
+              <targetTests>
+                <param>com.puppycrawl.tools.checkstyle.xpath.*</param>
+              </targetTests>
+              <coverageThreshold>100</coverageThreshold>
+              <mutationThreshold>100</mutationThreshold>
+              <timeoutFactor>${pitest.plugin.timeout.factor}</timeoutFactor>
+              <timeoutConstant>${pitest.plugin.timeout.constant}</timeoutConstant>
+              <threads>${pitest.plugin.threads}</threads>
+            </configuration>
+          </plugin>
+        </plugins>
+      </build>
+    </profile>
   </profiles>
 </project>

```

File: .circleci/config.yml [hunk 1]
```diff
@@ -64,6 +64,7 @@ jobs:
               echo "./.ci/pitest.sh pitest-common"  >> commands.txt
               echo "./.ci/pitest.sh pitest-imports" >> commands.txt
               echo "./.ci/pitest.sh pitest-ant"     >> commands.txt
+              echo "./.ci/pitest.sh pitest-meta"    >> commands.txt
```
Comment (romani) [inline .circleci/config.yml:67]: execution is ignored .... as we has limit for 4 containers
please activate https://github.com/checkstyle/checkstyle/blob/master/.circleci/config.yml#L261
and put in `pitest7`

Comment (romani) [reply .circleci/config.yml:67]: as we ignore it and we know it is not finishing ...
please put in pitest7 but keep it commented out

Comment (gaurabdg) [reply .circleci/config.yml:67]: I think we should remove this line altogether, as we dont have `pitest-meta` in `pom.xml`

Comment (gaurabdg) [reply .circleci/config.yml]: Done

**Review by strkkk — APPROVED** [review]

**Pull request merged**